crates/matrix-sdk/src/event_cache/mod.rs

```diff
@@ -409,52 +415,65 @@ impl RoomEventCache {
-        &self,
-        max_wait: Option<Duration>,
-    ) -> Result<Option<PaginationToken>> {
-        self.inner.oldest_backpagination_token(max_wait).await
-    }
-
-    /// Back-paginate with the given token, if provided.
+    /// This automatically takes care of waiting for a pagination token from
+    /// sync, if we haven't done that before.
     ///
-    /// If no token has been provided, it will back-paginate from the end of the
-    /// room.
+    /// # Errors
     ///
-    /// If a token has been provided, but it was unknown to the event cache
-    /// (i.e. it's not associated to any gap in the timeline stored by the
-    /// event cache), then an error result will be returned.
-    pub async fn backpaginate(
-        &self,
-        batch_size: u16,
-        token: Option<PaginationToken>,
-    ) -> Result<BackPaginationOutcome> {
-        self.inner.backpaginate(batch_size, token).await
+    /// It may return an error if the pagination token used during
+    /// back-pagination has disappeared while we started the pagination. In
+    /// that case, it's desirable to call the method again.
+    ///
+    /// TODO(bnjbvr): that's bad API, this API should restart for us!
+    pub async fn paginate_backwards(&self, batch_size: u16) -> Result<BackPaginationOutcome> {
+        self.inner.paginate_backwards(batch_size).await
+    }
+
+    /// Test-only function to get the latest pagination token, as stored in the
+    /// room events linked list.
+    #[cfg(any(test, feature = "testing"))]
+    #[doc(hidden)]
+    pub async fn get_or_wait_for_pagination_token(&self) -> Option<String> {
+        self.inner.get_or_wait_for_pagination_token().await
+    }
+
+    /// Returns a subscriber to the pagination status used for the
+    /// back-pagination integrated to the event cache.
+    pub fn pagination_status(&self) -> Subscriber<PaginatorState> {
+        self.inner.pagination.paginator.state()
     }
 }
 
+#[derive(Debug)]
+struct RoomPaginationData {
+    /// The stateful paginator instance used for the integrated pagination.
+    paginator: Paginator,
+
+    /// Have we ever waited for a previous-batch-token from the sync? We do this
+    /// at most once per room, and we reset that upon clearing the timeline
+    /// events.
+    waited_for_initial_prev_token: Mutex<bool>,
+}
+
 /// The (non-clonable) details of the `RoomEventCache`.
 struct RoomEventCacheInner {
     /// Sender part for subscribers to this room.
     sender: Sender<RoomEventCacheUpdate>,
 
-    /// The Client [`Room`] this event cache pertains to.
-    room: Room,
-
     /// The events of the room.
     events: RwLock<RoomEvents>,
 
     /// A notifier that we received a new pagination token.
     pagination_token_notifier: Notify,
 
-    /// A lock that ensures we don't run multiple pagination queries at the same
-    /// time.
-    pagination_lock: Mutex<()>,
+    /// A paginator instance, that's configured to run back-pagination on our
```

Comment:
It doesn't handle front-pagination? Why?

Reply:
Expanded the comment, but TLDR; this is only for back-paginations that interact with the event cache. Forward paginations don't, at the moment (i.e. events obtained in a response of forward pagination will NOT be saved into the cache — such paginations are "out of band")